### ISSUE TITLE
Downloads dropping to zero bytes/sec

### ISSUE THREAD

**Issue by synctext** [Tribler/tribler#3379]
To reproduce:
- join a big swarm
- see connections grow to 125+ connections
- Tribler downloads drop to 0 bytes/sec slowly
- GUI is still repsonding
- Tribler re-start does not work
- hanging background process of Tribler needs a kill -9, but that not always works 
- reboot your Mac

**Comment by devos50** [comment]
We haven't been able to accurately reproduce this so far. Let's postpone it to 7.1.

**Comment by Dmole** [comment]
Maybe related; The thread count (as reported by Activity Monitor) increases up to ~1000 threads for ~1 peer then drops down to ~500 or ~200, and repeats.
Only on the lower PID (I think that's the UI process).
Debug>System>threads only lists 14 though (likely some of the 26 in the core process).

**Comment by synctext** [comment]
Brainstorming on a long-term solution... Having a list of 100 magnet links of http://torrent.ubuntu.com/releases/artful/release/server/, or http://bt.etree.org/index.php?sort=seeders 
then see if this breaks things. Just download them _all_ 100 concurrently.

**Comment by synctext** [comment]
Reproduced... Now shows the core not responding or something.
<img width="1280" alt="screen shot 2018-03-28 at 17 11 39" src="https://user-images.githubusercontent.com/325224/38039330-519cc76c-32ad-11e8-96d1-6acae374c65d.png">

**Comment by synctext** [comment]
Another observation using standard ```top```. Normally this Mac has lots of threads running, in _nothing running_ state. The Tribler core process below only seems to use 21 _#ports_. When the Tribler core is 90% in the ```stuck``` state, it's clear we're in trouble. Thus we have an alternative for a continuous keep-alive ping to check the core.
<img width="1280" alt="screen shot 2018-03-30 at 00 04 00" src="https://user-images.githubusercontent.com/325224/38131048-bfbfec52-3405-11e8-9b65-e0fa7697ca3d.png">

**Comment by devos50** [comment]
@xoriole @synctext is this issue still valid?

**Comment by devos50** [comment]
I think I managed to (somewhat) reproduce this with the application tester. After running the application tester for 16 hours on our Mac, I observed the following:

![speed_down](https://user-images.githubusercontent.com/1707075/44137576-34ec11f4-a071-11e8-8b92-add1ac116d57.png)

Note that it could also be because the downloads are completed and the application tester is not removing them correctly. Unfortunately, I didn't monitor the progress of the downloads so I extend the application tester to do so.

**Comment by devos50** [comment]
The next step to debug this issue, as discussed yesterday, is to download some torrents without anonymity and plot the same graph. This requires some changes to the application tester.

**Comment by synctext** [comment]
Mac performance of [7.1.0 RC1 release](https://github.com/Tribler/tribler/releases/tag/v7.1.0-rc1) is very impressive. Trying to crash it, but did not succeed yet. CPU usage basically collapsed by upgrading from the Beta to RC1! Tribler no longer consumes the majority of a CPU core. Old Mac laptop did not even kick in the fan. This has never happened. Next is to even remove the 50 max. socket connections in the settings. See direct download in 3 large swarms. Please continue with the application tester, we need more soak testing and multi-platform test coverage.

***After 13 years of trying, it seems that we finally have something that is becoming usable.***
<img width="619" alt="screen shot 2018-09-12 at 22 13 55" src="https://user-images.githubusercontent.com/325224/45686081-bf3a2400-bb4b-11e8-8658-2944e6460707.png">

**Comment by synctext** [comment]
This is causing trouble for our users. Escalating this to top-priority, as it severly impact usability.
https://forum.tribler.org/t/tribler-starts-up-finds-seeders-starts-downloading-and-after-2-minutes-stops-and-all-connections-are-down-to-0/4247/2

**Comment by xoriole** [comment]
Some observations:
- The number of connected peers reduces to zero which is why speed is zero
- Not all torrents have the issue at the same time. Other torrents could be downloading fine.
- Issue not yet observed(or reproduced) on torrents added from torrent file with trackers. The issue is highly visible on magnet links torrents with no trackers listed.

**Comment by synctext** [comment]
No, bug is when connected to 50+ peers. Running smooth, then sudden drop to speed of 0.

**Comment by xoriole** [comment]
Yes, GUI does show a high number of seeds/peers but the actual libtorrent peers slowly reduces to zero. An example of this case is shown in this screenshot below.
![large-swarm](https://user-images.githubusercontent.com/1442867/47015577-a70af400-d14d-11e8-9d55-a713c756dadc.png)

**Comment by devos50** [comment]
@Captain-Coder also noticed this bug; he will also try to reproduce/investigate it 👍

**Comment by Dmole** [comment]
synctext, xoriole is correct the 349 peer count is # of peers found, the list of 0 is the for currently connected peers. I listed some mitigating features here; https://github.com/Tribler/tribler/issues/3857

**Comment by synctext** [comment]
Progress! Tested custom build. TCP-only is better. Still very eradic speeds. This might be a case of CPU hogging. Then utp packet get delivered late, measured with high latency and agressive throttle kicks in. Please try to reproduce on Jenkins at some point.

**Comment by devos50** [comment]
Interesting that TCP-only works better. At this point, it might be helpful to make a libtorrent-only package (with PyInstaller), start a single download using that library, and compare the single-download speed to that of a download in Tribler. This could reveal whether the problem is in libtorrent or Tribler (I suspect the latter).

**Comment by synctext** [comment]
ToDo: please create infrastructure to reproduce. Have all our 1 Gbps servers and few laptops seed a 5GB swarm. Both 1 swarm with no trackers and 1 swarm with trackers.

**Comment by devos50** [comment]
Created a minimal libtorrent application to hopefully reproduce this bug (code can be found [here](https://github.com/devos50/libtorrent-python-minimal)). The binaries can be found on Jenkins ([MacOS](https://jenkins-ci.tribler.org/job/pers/job/libtorrent_minimal_mac/) and [Windows 64-bit](https://jenkins-ci.tribler.org/job/pers/job/libtorrent_minimal_win64/)). I've started this application on our Mac Tester, running for 1.5h+ now without a drop in download speed.

TODO: make sure that the libtorrent session configuration is as close as possible to the Tribler one.

**Comment by qstokkink** [comment]
Has been running about 10 minutes on Windows before freeze.

**Comment by synctext** [comment]
What is dmg or something in these files? Libtorrentminimal.dms has issues

**Comment by synctext** [comment]
Running the app. Not stuck, but very sensitive to wifi changes. Bizar! Bandwidth collapses if you move the laptop, even if you still have 4 full signal strength bars. Recovers in 20-30 seconds.

**Comment by devos50** [comment]
Hmm that's interesting. Do you see the console output? I'm printing alerts of the `error` category, but I will change the app so we can see the `performance` alerts. It might help us to get a bit more insight in what's going on.

Today, the app was running for over 4 hours without any problems on the Mac tester (i.e. downloads never dropped to zero for unknown reasons, except when the download got removed and started again). One thing I would like to add, is that it writes the download speed to a separate file, which can then be plotted [with this R script](https://github.com/Tribler/application-tester/blob/master/scripts/downloads.r). Next, run it on all platforms we have, for prolonged periods of time. Also, we need to bundle the `.torrent` files since the request to `http://releases.ubuntu.com` sometimes fails, resulting in a download state we cannot recover from.

**Comment by devos50** [comment]
Also, running the application is a bit non-trivial. You should download the archive from Jenkins (which includes the `.app` file), navigate in the `.app` file itself, go to the `MacOS` directory and run the `libtorrentminimal` binary from the terminal.

**Comment by synctext** [comment]
Downloaded and reproduced this with latest v7.1.2 version of Tribler. As requested I've downloaded two files and tried to see the effect when reproducing the error. Standard Ubuntu download with plenty of seeds has a "bandwidth crash" from 5.0 MByte/sec to zero. First the health mode:

![screenshot_20181029-204651](https://user-images.githubusercontent.com/325224/47676372-c75e9800-dbbc-11e8-8a29-0f7defe1dca5.png)

Then a live capture of the after the crash kicks in (did not manage to capture the crash itself):

https://photos.app.goo.gl/DZau4bevHH7WPt819

**Comment by xoriole** [comment]
It is different from what I've observed so far. In my case, when the download speed was dropping, the number of connected seeders was dropping with it and there were zero connected peers when download speed became 0kB/s. But what you have is quite different and interesting. The seeders are still connected but not contributing. This gives a feeling that this Tribler peer is being choked by all the connected seeders at the same time which I think is less likely.

**Comment by synctext** [comment]
Finally reproduced!
Seems downloading to USB fails.
After the developers dived deeper it seems even naked Libtorrent stalls when writing to an external USB stick. Facinating.

**Comment by ichorid** [comment]
Seems like most USB drives have a real problem with random writes:
![usb](https://user-images.githubusercontent.com/2509103/47926586-3ab82080-dec1-11e8-96d0-2a9c6b6953a5.png)

**Comment by synctext** [comment]
if you understand when it fails, Tribler is amazing..
Conducted an experiment to understand if the USB bottleneck theory holds. It is a Windows-compatible large format USB drive, mounted on Mac. In the past I've seen that on Mac the shutdown can take over 30minutes. My understanding was that Tribler has gone in some error state, went corrupt, ran out of file descriptors, something. Nope, it's merely 30minutes of backlog USB stick pending writes; it seems. Experiment:
- download on fast internal harddisk; stable download
- start second download to slow USB stick
- watch as both downloads experience a "bandwidth crash".
- stop the download to the USB stick
- wait for 30 seconds to see the other download jump to 1 Kbps, then crash to 0 again.
- wait longer, download resumes, finishes
- Tribler then closes responsively within seconds!

@ichorid @xoriole Please try to make a reproducible _slow IO_ testing ecosystems for this within Jenkins/nosetest. Do we need something complex like ```ionice``` or [trickle](https://github.com/mariusae/trickle)+[NBD](https://sweetcode.io/introduction-to-linux-network-block-devices/)?

**Comment by devos50** [comment]
@Captain-Coder this is a very interesting observation!

So there are several things here that might lead to this issue. From the log:

```
2019-01-15T13:20:31+0000 [stderr#error] INFO:TriblerTunnelCommunity:Want 0 data circuits of length 1
2019-01-15T13:20:31+0000 [stderr#error] INFO:TriblerTunnelCommunity:Removing circuit 1682128877 too old
2019-01-15T13:20:32+0000 [stderr#error] INFO:TriblerTunnelCommunity:Removed circuit 1682128877 too old
```

So it is removing a circuit, and pops it from `self.circuits`. Five seconds later, the `do_circuits` is called again and it should create a new circuit. The log says:

```
2019-01-15T13:20:36+0000 [stderr#error] INFO:TriblerTunnelCommunity:Want 0 data circuits of length 1
```

So if we look at the code in the `do_circuits` method, it implies that either `settings.max_circuits` <= 8 or we have at least 8 circuits already.

I suspect that the DHT lookups are messing with a counter or circuit creation?

**Comment by synctext** [comment]
"TriblerTunnelCommunity keeps telling us it wants zero tunnels"
Thnx for insight into this top-priority bug! Great use-case for soak testing e2e anon downloads.

**Comment by Captain-Coder** [comment]
I added a debug statement to do_circuits. It shows circuits_needed going to 0. There is only one place in the code where this assignment is made triblertunnel/community.py:on_download_removed(). But it is only made 0 when num_hops_by_downloads goes to 0.

I use Tribler headless. So I add downloads by bash scripts & curl, and I have a systemd timer that runs a check for completed downloads and removes them. And the problem occured some 10 minutes after a completed download was removed and the last tunnels have expired. So that coincides with the on_download_removed() code.

So how does the num_hops_by_downloads counter work? It should count the number of active downloads for a given number of hops. It gets reduced when a download is removed. And it gets incremented... only once. This happens in build_tunnels(). Which in turn gets called only once, from deep within can_create_engine_wrapper(). Yes that big mess.

The can_create_engine_wrapper() first checks if there is a need to build tunnels, and then calls build_tunnels(). If there are preexisting tunnels (like a download already going) build_tunnels() will be skipped, num_hops_by_downloads won't be incremented and is no longer accurate. This in turn causes the circuits_needed to go to 0 when a download is removed.

The obvious flaw is that build_tunnels should not be expected to be called exactly once per download. This begs the question of why build_tunnels is bothering to keep track of downloads anyway, it is in the ipv8 part of the tunnel community. The only reason why num_hops_by_downloads is there is for the tunnels_ready() function, which probaly should look to circuits_needed and not num_hops_by_downloads. I think that this could be resolved by moving num_hops_by_downloads to triblertunnel/community.py and finding a suitable place for the increment, or not bothering with keeping track exactly and just set it as a whole in the monitor_downloads() callback. I'd do this, but that is a PR over 2 repos, I have no time and I need to write experiments and text.
